In [197]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns 
%matplotlib inline

In [198]:
claim=pd.read_csv('Claim.csv')
Demographics=pd.read_csv('Demographics.csv')
Fraud_data=pd.read_csv('Fraud Data.csv')
Policy=pd.read_csv('Policy.csv')
Vehicle=pd.read_csv('Vehicle.csv')

In [199]:
claim

CustomerID DateOfIncident            TypeOfIncident TypeOfCollission  \
0      Cust10000     2015-02-03   Multi-vehicle Collision   Side Collision   
1      Cust10001     2015-02-02   Multi-vehicle Collision   Side Collision   
2      Cust10002     2015-01-15  Single Vehicle Collision   Side Collision   
3      Cust10003     2015-01-19  Single Vehicle Collision   Side Collision   
4      Cust10004     2015-01-09  Single Vehicle Collision   Rear Collision   
...          ...            ...                       ...              ...   
28831   Cust9993     2015-01-24             Vehicle Theft                ?   
28832   Cust9994     2015-02-09  Single Vehicle Collision  Front Collision   
28833   Cust9996     2015-01-28  Single Vehicle Collision   Rear Collision   
28834   Cust9997     2015-01-28  Single Vehicle Collision   Rear Collision   
28835   Cust9999     2015-01-13  Single Vehicle Collision  Front Collision   

      SeverityOfIncident AuthoritiesContacted IncidentState IncidentCity  \
0             Total Loss               Police        State7        City1   
1             Total Loss               Police        State7        City5   
2           Minor Damage                Other        State8        City6   
3           Minor Damage                Other        State9        City6   
4           Minor Damage                 Fire        State8        City6   
...                  ...                  ...           ...          ...   
28831     Trivial Damage               Police        State9        City4   
28832       Minor Damage                 Fire        State8        City3   
28833       Minor Damage                 Fire        State9        City7   
28834       Minor Damage            Ambulance        State9        City3   
28835         Total Loss                Other        State8        City3   

      IncidentAddress  IncidentTime  NumberOfVehicles PropertyDamage  \
0       Location 1311            17                 3              ?   
1       Location 1311            10                 3            YES   
2       Location 2081            22                 1            YES   
3       Location 2081            22                 1            YES   
4       Location 1695            10                 1             NO   
...               ...           ...               ...            ...   
28831   Location 1890             3                 1             NO   
28832   Location 2097            17                 1            YES   
28833   Location 1452             1                 1            YES   
28834   Location 1876             1                 1              ?   
28835   Location 1874             9                 1             NO   

       BodilyInjuries Witnesses PoliceReport AmountOfTotalClaim  \
0                   1         0            ?              65501   
1                   2         1          YES              61382   
2                   2         3           NO              66755   
3                   2         3           NO              66243   
4                   2         1          YES              53544   
...               ...       ...          ...                ...   
28831               0         3            ?               6590   
28832               0         2          YES              74547   
28833               2         0          YES              55305   
28834               2         0          YES              55830   
28835               1         3          YES              68969   

       AmountOfInjuryClaim  AmountOfPropertyClaim  AmountOfVehicleDamage  
0                    13417                   6071                  46013  
1                    15560                   5919                  39903  
2                    11630                  11630                  43495  
3                    12003                  12003                  42237  
4                     8829                   7234                  37481  
...                    ...          

In [200]:
Demographics

CustomerID  InsuredAge  InsuredZipCode InsuredGender  \
0      Cust10000          35          454776          MALE   
1      Cust10001          36          454776          MALE   
2      Cust10002          33          603260          MALE   
3      Cust10003          36          474848          MALE   
4      Cust10004          29          457942        FEMALE   
...          ...         ...             ...           ...   
28831   Cust9993          46          600561          MALE   
28832   Cust9994          44          439304          MALE   
28833   Cust9996          53          460722          MALE   
28834   Cust9997          53          472634          MALE   
28835   Cust9999          36          450730        FEMALE   

      InsuredEducationLevel  InsuredOccupation InsuredHobbies  CapitalGains  \
0                        JD       armed-forces         movies         56700   
1                        JD       tech-support      cross-fit         70600   
2                        JD       armed-forces           polo         66400   
3                        JD       armed-forces           polo         47900   
4               High School    exec-managerial        dancing             0   
...                     ...                ...            ...           ...   
28831               Masters    protective-serv       sleeping             0   
28832               Masters       adm-clerical        dancing             0   
28833                   PhD   transport-moving   base-jumping         63100   
28834             Associate   transport-moving      skydiving         67400   
28835                   PhD  handlers-cleaners   base-jumping         46400   

       CapitalLoss Country  
0           -48500   India  
1           -48500   India  
2           -63700   India  
3           -73400   India  
4           -41500   India  
...            ...     ...  
28831            0   India  
28832       -28800   India  
28833       -43800   India  
28834       -43800   India  
28835       -74300   India  

[28836 rows x 10 columns]

In [201]:
Fraud_data

CustomerID ReportedFraud
0      Cust20065             N
1      Cust37589             N
2      Cust24312             N
3       Cust5493             Y
4       Cust7704             Y
...          ...           ...
28831   Cust9773             N
28832  Cust38439             N
28833  Cust38444             N
28834  Cust26466             N
28835  Cust29555             N

[28836 rows x 2 columns]

In [202]:
Policy

InsurancePolicyNumber  CustomerLoyaltyPeriod DateOfPolicyCoverage  \
0                     110122                    328           2014-10-17   
1                     110125                    256           1990-05-25   
2                     110126                    228           2014-06-06   
3                     110127                    256           2006-10-12   
4                     110128                    137           2000-06-04   
...                      ...                    ...                  ...   
28831                 148614                    425           1999-10-12   
28832                 148615                    367           1996-11-09   
28833                 148616                    208           1995-12-04   
28834                 148618                    401           1994-09-12   
28835                 148619                    331           2006-07-16   

      InsurancePolicyState Policy_CombinedSingleLimit  Policy_Deductible  \
0                   State3                    250/500               1000   
1                   State1                    250/500               2000   
2                   State1                   500/1000               1000   
3                   State3                    250/500               1000   
4                   State2                    250/500               1000   
...                    ...                        ...                ...   
28831               State3                    250/500               1000   
28832               State3                    250/500               1000   
28833               State3                    250/500               1000   
28834               State3                    250/500                856   
28835               State3                    250/500               1000   

       PolicyAnnualPremium  UmbrellaLimit InsuredRelationship CustomerID  
0                  1406.91              0             husband   Cust1001  
1                  1415.74        6000000           unmarried   Cust1004  
2                  1583.91        6000000           unmarried   Cust1005  
3                  1351.10              0           unmarried   Cust1006  
4                  1333.35              0             husband   Cust1007  
...                    ...            ...                 ...        ...  
28831               839.81              0                wife  Cust39493  
28832               811.03              0             husband  Cust39494  
28833               891.37              0             husband  Cust39495  
28834               911.37              0           own-child  Cust39497  
28835              1062.91              0                wife  Cust39498  

[28836 rows x 10 columns]

In [203]:
Vehicle

CustomerID VehicleAttribute VehicleAttributeDetails
0       Cust20179        VehicleID             Vehicle8898
1       Cust21384     VehicleModel                  Malibu
2       Cust33335      VehicleMake                  Toyota
3       Cust27118     VehicleModel                    Neon
4       Cust13038        VehicleID            Vehicle30212
...           ...              ...                     ...
115339  Cust22134      VehicleMake                    Audi
115340  Cust25338        VehicleID            Vehicle28831
115341   Cust7637        VehicleID            Vehicle12870
115342  Cust33581        VehicleID            Vehicle19680
115343  Cust14156     VehicleModel                 Corolla

[115344 rows x 3 columns]

In [204]:
pd.DataFrame(Vehicle).VehicleAttribute.unique()

array(['VehicleID', 'VehicleModel', 'VehicleMake', 'VehicleYOM'],
      dtype=object)

In [205]:
newdf1 = Vehicle[(Vehicle['VehicleAttribute']=='VehicleID')].drop(['VehicleAttribute'],axis=1).rename(columns = {'VehicleAttributeDetails':'VehicleID'})
newdf2 = Vehicle[(Vehicle['VehicleAttribute']=='VehicleModel')].drop(['VehicleAttribute'],axis=1).rename(columns = {'VehicleAttributeDetails':'VehicleModel'})
newdf3 = Vehicle[(Vehicle['VehicleAttribute']=='VehicleMake')].drop(['VehicleAttribute'],axis=1).rename(columns = {'VehicleAttributeDetails':'VehicleMake'})
newdf4 = Vehicle[(Vehicle['VehicleAttribute']=='VehicleYOM')].drop(['VehicleAttribute'],axis=1).rename(columns = {'VehicleAttributeDetails':'VehicleYOM'})

In [206]:
Vehicledata=newdf1.merge(newdf2,how='outer',on='CustomerID').merge(newdf3,how='outer',on='CustomerID').merge(newdf4,how='outer',on='CustomerID')

In [207]:
Vehicledata

CustomerID     VehicleID VehicleModel VehicleMake VehicleYOM
0      Cust20179   Vehicle8898    Forrestor         BMW       2011
1      Cust13038  Vehicle30212       Maxima      Toyota       2000
2       Cust1801  Vehicle24096          MDX      Accura       1999
3      Cust14947  Vehicle15216       Ultima      Nissan       1998
4      Cust19144  Vehicle29018       Fusion        Ford       2011
...          ...           ...          ...         ...        ...
28831   Cust2970  Vehicle19228   Pathfinder        Audi       2003
28832  Cust19731  Vehicle12644       Malibu   Chevrolet       2014
28833  Cust25338  Vehicle28831           95        Saab       2011
28834   Cust7637  Vehicle12870       Fusion        Ford       2008
28835  Cust33581  Vehicle19680     3 Series         BMW       2004

[28836 rows x 5 columns]

In [208]:
datamerge=Demographics.merge(Policy,how='outer',on='CustomerID').merge(claim,how='outer',on='CustomerID').merge(Vehicledata,how='outer',on='CustomerID').merge(Fraud_data,how='outer',on='CustomerID')

In [209]:
missing_values=['?',-5,'MISSINGVALUE','MISSEDDATA','NA',-1,'???']
datamerge=datamerge.replace(missing_values,np.NaN)
datamerge

CustomerID  InsuredAge  InsuredZipCode InsuredGender  \
0      Cust10000          35          454776          MALE   
1      Cust10001          36          454776          MALE   
2      Cust10002          33          603260          MALE   
3      Cust10003          36          474848          MALE   
4      Cust10004          29          457942        FEMALE   
...          ...         ...             ...           ...   
28831   Cust9993          46          600561          MALE   
28832   Cust9994          44          439304          MALE   
28833   Cust9996          53          460722          MALE   
28834   Cust9997          53          472634          MALE   
28835   Cust9999          36          450730        FEMALE   

      InsuredEducationLevel  InsuredOccupation InsuredHobbies  CapitalGains  \
0                        JD       armed-forces         movies         56700   
1                        JD       tech-support      cross-fit         70600   
2                        JD       armed-forces           polo         66400   
3                        JD       armed-forces           polo         47900   
4               High School    exec-managerial        dancing             0   
...                     ...                ...            ...           ...   
28831               Masters    protective-serv       sleeping             0   
28832               Masters       adm-clerical        dancing             0   
28833                   PhD   transport-moving   base-jumping         63100   
28834             Associate   transport-moving      skydiving         67400   
28835                   PhD  handlers-cleaners   base-jumping         46400   

       CapitalLoss Country  ...  PoliceReport  AmountOfTotalClaim  \
0           -48500   India  ...           NaN               65501   
1           -48500   India  ...           YES               61382   
2           -63700   India  ...            NO               66755   
3           -73400   India  ...            NO               66243   
4           -41500   India  ...           YES               53544   
...            ...     ...  ...           ...                 ...   
28831            0   India  ...           NaN                6590   
28832       -28800   India  ...           YES               74547   
28833       -43800   India  ...           YES               55305   
28834       -43800   India  ...           YES               55830   
28835       -74300   India  ...           YES               68969   

      AmountOfInjuryClaim AmountOfPropertyClaim AmountOfVehicleDamage  \
0                   13417                  6071                 46013   
1                   15560                  5919                 39903   
2                   11630                 11630                 43495   
3                   12003                 12003                 42237   
4                    8829                  7234                 37481   
...                   ...                   ...                   ...   
28831                 870                  1078                  4642   
28832               14699                  7875                 51973   
28833                7043                  7043                 41219   
28834                7059                  7059                 41712   
28835               12075                  6038                 50856   

          VehicleID  VehicleModel  VehicleMake VehicleYOM ReportedFraud  
0      Vehicle26917            A5         Audi       2008             N  
1      Vehicle15893            A5         Audi       2006             N  
2       Vehicle5152         Jetta   Volkswagen       1999             N  
3      Vehicle37363         Jetta   Volkswagen       2003             N  
4      Vehicle28633           CRV       Toyota       2010             N  
...             ...           ...          ...        ...           ...  
28831  Vehicle13568       Impreza       Suburu       2007             N  
28832  Vehicle

In [210]:
datamerge=datamerge.drop (['Country','CustomerID','DateOfPolicyCoverage','DateOfIncident','IncidentAddress','VehicleID'],axis=1)
datamerge

InsuredAge  InsuredZipCode InsuredGender InsuredEducationLevel  \
0              35          454776          MALE                    JD   
1              36          454776          MALE                    JD   
2              33          603260          MALE                    JD   
3              36          474848          MALE                    JD   
4              29          457942        FEMALE           High School   
...           ...             ...           ...                   ...   
28831          46          600561          MALE               Masters   
28832          44          439304          MALE               Masters   
28833          53          460722          MALE                   PhD   
28834          53          472634          MALE             Associate   
28835          36          450730        FEMALE                   PhD   

       InsuredOccupation InsuredHobbies  CapitalGains  CapitalLoss  \
0           armed-forces         movies         56700       -48500   
1           tech-support      cross-fit         70600       -48500   
2           armed-forces           polo         66400       -63700   
3           armed-forces           polo         47900       -73400   
4        exec-managerial        dancing             0       -41500   
...                  ...            ...           ...          ...   
28831    protective-serv       sleeping             0            0   
28832       adm-clerical        dancing             0       -28800   
28833   transport-moving   base-jumping         63100       -43800   
28834   transport-moving      skydiving         67400       -43800   
28835  handlers-cleaners   base-jumping         46400       -74300   

       InsurancePolicyNumber  CustomerLoyaltyPeriod  ... Witnesses  \
0                     119121                     49  ...         0   
1                     119122                    114  ...         1   
2                     119123                    167  ...         3   
3                     119124                    190  ...         3   
4                     119125                    115  ...         1   
...                      ...                    ...  ...       ...   
28831                 119114                    321  ...         3   
28832                 119115                    237  ...         2   
28833                 119117                    392  ...         0   
28834                 119118                    391  ...         0   
28835                 119120                    191  ...         3   

      PoliceReport  AmountOfTotalClaim  AmountOfInjuryClaim  \
0              NaN               65501                13417   
1              YES               61382                15560   
2               NO               66755                11630   
3               NO               66243                12003   
4              YES               53544                 8829   
...            ...                 ...                  ...   
28831          NaN                6590                  870   
28832          YES               74547                14699   
28833          YES               55305                 7043   
28834          YES               55830                 7059   
28835          YES               68969                12075   

       AmountOfPropertyClaim AmountOfVehicleDamage VehicleModel VehicleMake  \
0                       6071                 46013           A5        Audi   
1                       5919                 39903           A5        Audi   
2                      11630                 43495        Jetta  Volkswagen   
3                      12003                 42237        Jetta  Volkswagen   
4                       7234                 37481          CRV      Toyota   
...                      ...                   ...          ...         ...   
28831                   1078                  4642      Impreza      Suburu   
28832                   7875                 51973     Wran

In [211]:
datamerge['InsuredGender'] = datamerge['InsuredGender'].fillna(datamerge['InsuredGender'].mode()[0])
datamerge['PolicyAnnualPremium'] = datamerge['PolicyAnnualPremium'].fillna(datamerge['PolicyAnnualPremium'].mode()[0])
datamerge['TypeOfCollission'] = datamerge['TypeOfCollission'].fillna(datamerge['TypeOfCollission'].mode()[0])
datamerge['IncidentTime'] = datamerge['IncidentTime'].fillna(datamerge['IncidentTime'].median())
datamerge['PropertyDamage'] = datamerge['PropertyDamage'].fillna(datamerge['PropertyDamage'].mode()[0])
datamerge['Witnesses'] = datamerge['Witnesses'].fillna(datamerge['Witnesses'].median())
datamerge['PoliceReport'] = datamerge['PoliceReport'].fillna(datamerge['PoliceReport'].mode()[0])
datamerge['AmountOfTotalClaim'] = datamerge['AmountOfTotalClaim'].fillna(datamerge['AmountOfTotalClaim'].median())
datamerge['VehicleMake'] = datamerge['VehicleMake'].fillna(datamerge['VehicleMake'].mode()[0])

In [212]:
datamerge

InsuredAge  InsuredZipCode InsuredGender InsuredEducationLevel  \
0              35          454776          MALE                    JD   
1              36          454776          MALE                    JD   
2              33          603260          MALE                    JD   
3              36          474848          MALE                    JD   
4              29          457942        FEMALE           High School   
...           ...             ...           ...                   ...   
28831          46          600561          MALE               Masters   
28832          44          439304          MALE               Masters   
28833          53          460722          MALE                   PhD   
28834          53          472634          MALE             Associate   
28835          36          450730        FEMALE                   PhD   

       InsuredOccupation InsuredHobbies  CapitalGains  CapitalLoss  \
0           armed-forces         movies         56700       -48500   
1           tech-support      cross-fit         70600       -48500   
2           armed-forces           polo         66400       -63700   
3           armed-forces           polo         47900       -73400   
4        exec-managerial        dancing             0       -41500   
...                  ...            ...           ...          ...   
28831    protective-serv       sleeping             0            0   
28832       adm-clerical        dancing             0       -28800   
28833   transport-moving   base-jumping         63100       -43800   
28834   transport-moving      skydiving         67400       -43800   
28835  handlers-cleaners   base-jumping         46400       -74300   

       InsurancePolicyNumber  CustomerLoyaltyPeriod  ... Witnesses  \
0                     119121                     49  ...         0   
1                     119122                    114  ...         1   
2                     119123                    167  ...         3   
3                     119124                    190  ...         3   
4                     119125                    115  ...         1   
...                      ...                    ...  ...       ...   
28831                 119114                    321  ...         3   
28832                 119115                    237  ...         2   
28833                 119117                    392  ...         0   
28834                 119118                    391  ...         0   
28835                 119120                    191  ...         3   

      PoliceReport  AmountOfTotalClaim  AmountOfInjuryClaim  \
0               NO               65501                13417   
1              YES               61382                15560   
2               NO               66755                11630   
3               NO               66243                12003   
4              YES               53544                 8829   
...            ...                 ...                  ...   
28831           NO                6590                  870   
28832          YES               74547                14699   
28833          YES               55305                 7043   
28834          YES               55830                 7059   
28835          YES               68969                12075   

       AmountOfPropertyClaim AmountOfVehicleDamage VehicleModel VehicleMake  \
0                       6071                 46013           A5        Audi   
1                       5919                 39903           A5        Audi   
2                      11630                 43495        Jetta  Volkswagen   
3                      12003                 42237        Jetta  Volkswagen   
4                       7234                 37481          CRV      Toyota   
...                      ...                   ...          ...         ...   
28831                   1078                  4642      Impreza      Suburu   
28832                   7875                 51973     Wran

In [213]:
datamerge.isna().sum()

InsuredAge                    0
InsuredZipCode                0
InsuredGender                 0
InsuredEducationLevel         0
InsuredOccupation             0
InsuredHobbies                0
CapitalGains                  0
CapitalLoss                   0
InsurancePolicyNumber         0
CustomerLoyaltyPeriod         0
InsurancePolicyState          0
Policy_CombinedSingleLimit    0
Policy_Deductible             0
PolicyAnnualPremium           0
UmbrellaLimit                 0
InsuredRelationship           0
TypeOfIncident                0
TypeOfCollission              0
SeverityOfIncident            0
AuthoritiesContacted          0
IncidentState                 0
IncidentCity                  0
IncidentTime                  0
NumberOfVehicles              0
PropertyDamage                0
BodilyInjuries                0
Witnesses                     0
PoliceReport                  0
AmountOfTotalClaim            0
AmountOfInjuryClaim           0
AmountOfPropertyClaim         0
AmountOf

In [214]:
datamerge['VehicleYOM'] = pd.to_numeric(datamerge['VehicleYOM'])
datamerge['AmountOfTotalClaim'] = pd.to_numeric(datamerge['AmountOfTotalClaim'])
datamerge['Witnesses'] = pd.to_numeric(datamerge['Witnesses'])
datamerge['ReportedFraud'] = datamerge['ReportedFraud'].map({'Y': 1, 'N': 0})

In [215]:
datamerge['ReportedFraud'].value_counts()

0    21051
1     7785
Name: ReportedFraud, dtype: int64

In [216]:
dummies=['InsuredGender','InsuredEducationLevel','InsuredOccupation','InsuredHobbies','InsurancePolicyState','Policy_CombinedSingleLimit','InsuredRelationship','TypeOfIncident','TypeOfCollission','SeverityOfIncident','AuthoritiesContacted','IncidentState','IncidentCity','PropertyDamage','PoliceReport','VehicleModel','VehicleMake']

In [217]:
final_data = pd.get_dummies(datamerge,columns=dummies,drop_first=True)
final_data

InsuredAge  InsuredZipCode  CapitalGains  CapitalLoss  \
0              35          454776         56700       -48500   
1              36          454776         70600       -48500   
2              33          603260         66400       -63700   
3              36          474848         47900       -73400   
4              29          457942             0       -41500   
...           ...             ...           ...          ...   
28831          46          600561             0            0   
28832          44          439304             0       -28800   
28833          53          460722         63100       -43800   
28834          53          472634         67400       -43800   
28835          36          450730         46400       -74300   

       InsurancePolicyNumber  CustomerLoyaltyPeriod  Policy_Deductible  \
0                     119121                     49               1000   
1                     119122                    114               1000   
2                     119123                    167                617   
3                     119124                    190                722   
4                     119125                    115                500   
...                      ...                    ...                ...   
28831                 119114                    321                655   
28832                 119115                    237               1089   
28833                 119117                    392                787   
28834                 119118                    391                780   
28835                 119120                    191               2000   

       PolicyAnnualPremium  UmbrellaLimit  IncidentTime  ...  \
0                  1632.73              0          17.0  ...   
1                  1255.19              0          10.0  ...   
2                  1373.38              0          22.0  ...   
3                  1337.60              0          22.0  ...   
4                  1353.73        4279863          10.0  ...   
...                    ...            ...           ...  ...   
28831              1276.01              0           3.0  ...   
28832              1273.38              0          17.0  ...   
28833              1380.92        3448735           1.0  ...   
28834              1389.29        3364301           1.0  ...   
28835               928.43        2909175           9.0  ...   

       VehicleMake_Dodge  VehicleMake_Ford  VehicleMake_Honda  \
0                      0                 0                  0   
1                      0                 0                  0   
2                      0                 0                  0   
3                      0                 0                  0   
4                      0                 0                  0   
...                  ...               ...                ...   
28831                  0                 0                  0   
28832                  0                 0                  0   
28833                  0                 0                  0   
28834                  0                 0                  0   
28835                  0                 0                  0   

       VehicleMake_Jeep  VehicleMake_Mercedes  VehicleMake_Nissan  \
0                     0                     0                   0   
1                     0                     0                   0   
2                     0                     0                   0   
3                     0                     0                   0   
4                     0                     0                   0   
...                 ...                   ...                 ...   
28831                 0                     0                   0   
28832                 1                     0                   0   
28833                 0                     0                   0   
28834                 0                     0                   0   
28835                 0                 

In [218]:
final_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 28836 entries, 0 to 28835
Columns: 150 entries, InsuredAge to VehicleMake_Volkswagen
dtypes: float64(4), int64(15), uint8(131)
memory usage: 8.0 MB


In [219]:
x=final_data.drop(['ReportedFraud'],axis=1)
y=final_data['ReportedFraud']
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.3,random_state=101)

# Linear Regression

In [220]:
from sklearn.linear_model import LinearRegression
lm=LinearRegression()

In [221]:
lm.fit(x_train,y_train)

LinearRegression()

In [222]:
predictions=lm.predict(x_test)
from sklearn import metrics
print('MAE:',metrics.mean_absolute_error(y_test,predictions))
print('MSE:',metrics.mean_squared_error(y_test,predictions))
print('RMSE:',np.sqrt(metrics.mean_squared_error(y_test,predictions)))
print('R2:',metrics.r2_score(y_test,predictions))

MAE: 0.2536272627967262
MSE: 0.11416878721199285
RMSE: 0.33788872016093235
R2: 0.4217274283212663


# Logistic Regression

In [223]:
from sklearn.linear_model import LogisticRegression
logmodel=LogisticRegression()
logmodel.fit(x_train,y_train)

C:\Users\Your\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [224]:
predictions=logmodel.predict(x_test)
from sklearn.metrics import classification_report
print(classification_report(y_test,predictions))

              precision    recall  f1-score   support

           0       0.77      0.94      0.85      6309
           1       0.60      0.22      0.32      2342

    accuracy                           0.75      8651
   macro avg       0.68      0.58      0.58      8651
weighted avg       0.72      0.75      0.70      8651



# DecisionTree Classifier

In [225]:
from sklearn.tree import DecisionTreeClassifier
dtree=DecisionTreeClassifier()

In [226]:
dtree.fit(x_train,y_train)

DecisionTreeClassifier()

In [227]:
predictions=dtree.predict(x_test)

In [228]:
print(classification_report(y_test,predictions))

              precision    recall  f1-score   support

           0       0.92      0.91      0.91      6309
           1       0.76      0.77      0.77      2342

    accuracy                           0.87      8651
   macro avg       0.84      0.84      0.84      8651
weighted avg       0.87      0.87      0.87      8651



# RandomForest Classifier

In [229]:
from sklearn.ensemble import RandomForestClassifier

In [230]:
rfc=RandomForestClassifier(n_estimators=500)

In [231]:
rfc.fit(x_train,y_train)

RandomForestClassifier(n_estimators=500)

In [232]:
predictions=rfc.predict(x_test)
print(classification_report(y_test,predictions))

              precision    recall  f1-score   support

           0       0.94      0.98      0.96      6309
           1       0.93      0.82      0.87      2342

    accuracy                           0.93      8651
   macro avg       0.93      0.90      0.91      8651
weighted avg       0.93      0.93      0.93      8651



# GaussianNB

In [233]:
from sklearn.naive_bayes import GaussianNB
NB_classifier=GaussianNB()

In [234]:
NB_classifier.fit(x_train,y_train)

GaussianNB()

In [235]:
y_predictions=NB_classifier.predict(x_test)
print(classification_report(y_test,y_predictions))

              precision    recall  f1-score   support

           0       0.86      0.79      0.82      6309
           1       0.53      0.65      0.58      2342

    accuracy                           0.75      8651
   macro avg       0.69      0.72      0.70      8651
weighted avg       0.77      0.75      0.76      8651



# SVC

In [236]:
from sklearn.svm import SVC
model=SVC()
model.fit(x_train,y_train)

SVC()

In [237]:
prediction=model.predict(x_test)
print(classification_report(y_test,prediction))

              precision    recall  f1-score   support

           0       0.73      1.00      0.84      6309
           1       0.00      0.00      0.00      2342

    accuracy                           0.73      8651
   macro avg       0.36      0.50      0.42      8651
weighted avg       0.53      0.73      0.62      8651



C:\Users\Your\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\Your\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\Your\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


# GradientBoosting Classifier

In [238]:
from sklearn.ensemble import GradientBoostingClassifier

In [239]:
xgb=GradientBoostingClassifier()
xgb.fit(x_train,y_train)

GradientBoostingClassifier()

In [240]:
prediction=xgb.predict(x_test)
print(classification_report(y_test,prediction))

              precision    recall  f1-score   support

           0       0.93      0.98      0.95      6309
           1       0.93      0.79      0.86      2342

    accuracy                           0.93      8651
   macro avg       0.93      0.89      0.90      8651
weighted avg       0.93      0.93      0.93      8651



# The best accuracy is 93%.Both RandomForest Classifier & GradientBoosting Classifier have the same accuracy